packages/fxa-payments-server/src/lib/flow-metrics.ts

```diff
@@ -0,0 +1,80 @@
+import SpeedTrap from 'speed-trap';
+import { Config } from './config';
+let initialized: boolean;
+const NOT_REPORTED_VALUE = 'none';
+
+export function init() {
+  if (!initialized) {
+    SpeedTrap.init();
+    initialized = true;
+  }
+}
+
+export function logLoadedEvent(
+  route: string,
+  config: Config,
+  getScreenInfo: Function
+) {
+  if (!initialized || !window.navigator.sendBeacon) {
+    return;
+  }
+
+  const loadData = SpeedTrap.getLoad();
+  const unloadData = SpeedTrap.getUnload();
+  const flowData = config.flowMetricsData;
+  const now = Date.now();
+  const screenInfo = getScreenInfo();
+
+  const metricsData = {
+    ...loadData,
+    ...unloadData,
+    ...flowData,
+    startTime: SpeedTrap.baseTime,
+    flushTime: now,
+    initialView: `payments.${route}`,
+    events: [
+      {
+        offset: now - SpeedTrap.baseTime,
+        type: 'loaded',
+      },
+    ],
+
+    // default / static / empty values
+    broker: NOT_REPORTED_VALUE,
+    client_id: NOT_REPORTED_VALUE,
+    context: 'web',
+    deviceId: NOT_REPORTED_VALUE,
+    entryPoint: NOT_REPORTED_VALUE,
+    entrypoint: NOT_REPORTED_VALUE,
+    experiments: [],
+    isSampledUser: false,
+    lang: config.lang,
+    marketing: [],
+    migration: NOT_REPORTED_VALUE,
+    referrer: window.document.referrer || NOT_REPORTED_VALUE,
+    screen: {
+      ...screenInfo,
+      height: screenInfo.screenHeight,
+      width: screenInfo.screenWidth,
+    },
+    service: 'payments-server',
+    uid: NOT_REPORTED_VALUE,
+    uniqueUserId: NOT_REPORTED_VALUE,
+    utm_campaign: NOT_REPORTED_VALUE,
+    utm_content: NOT_REPORTED_VALUE,
+    utm_medium: NOT_REPORTED_VALUE,
+    utm_source: NOT_REPORTED_VALUE,
+    utm_term: NOT_REPORTED_VALUE,
+  };
+
+  // This is not an Action insofar that it has no bearing on the app state.
+  window.navigator.sendBeacon(
+    `${config.servers.content.url}/metrics`,
```

Comment:
Looks like this PR uses the content server to submit metrics? I'm hoping to come up with an alternative in https://github.com/mozilla/fxa/issues/2425, not sure if this should land in the meantime. @shane-tomlinson thoughts?

Reply:
I suppose, what is the long term plan and is this approach on the path to that plan? I have wondered if reporting to the content server is a reasonable solution too, @philbooth pointed out we'd have to worry about naming collisions. What will these metrics look like when sent to re:dash and amplitude, will they be under their own namespace, or do we just depend on the service being 'payments-server'?

If sending to the content server, the [POST /metrics endpoint](https://github.com/mozilla/fxa/blob/1cb0d0a4e8104b700a78a414fcc2ecf96e0fd6fc/packages/fxa-content-server/server/lib/routes/post-metrics.js#L163) will need CORS enabled. Right now we only expect requests from the content server, which is on the same domain. To avoid displaying the full list of domains that are allowed to use CORS, we'd want to do something like [GET /metrics-flow](https://github.com/mozilla/fxa/blob/1cb0d0a4e8104b700a78a414fcc2ecf96e0fd6fc/packages/fxa-content-server/server/lib/routes/get-metrics-flow.js#L33), where the origin is compared against an allow list.

All that said, it would be worthwhile to explore what is needed to comple Phil's PR that adds a [POST /metrics endpoint](https://github.com/mozilla/fxa/pull/2734/files) to the payments server. There would be more Ops work to ensure those events make their way to Amplitude, but might be a reasonable tradeoff to simplify the mental model of how metrics work.

Reply:
This PR is strictly for flow performance metrics.  That is, `flow.performance.*` events written to `stderr`, then piped through https://github.com/mozilla/fxa-activity-metrics/, and resulting in https://sql.telemetry.mozilla.org/dashboard/fxa-performance.

That being said, there is nothing preventing us from collecting them in payments server.  I simply want to avoid duplication.

As for log pipeline, I suppose that isn't difficult per se?  I recently learned that the content server and payments server are on different cloud services.  But perhaps the logs are shipped to a central place first?  Anyways, it's probably pretty easy for Ops, but it's not nothing.

Namespace remains a (minor) concern as long as we collect these metrics into the same DB.  I've prefix'd `payments` to the metric type name, i.e. all the type names from this patch will begin with `flow.performance.payments`.

I thought CORS wasn't an issue since we were already allowing payments server to post CSP reports to content server.  But it doesn't sound like that's the case.  I shall make changes.

Reply:
@chenba I've sent you an invite to a payments metrics meeting Shane and I had scheduled for Thursday. We can figure out a plan then :+1: